File: module4_dr_grading python/src/generateDummyDataset.m
function generateDummyDataset(dataFolder, numImagesPerClass)
% GENERATEDUMMYDATASET Generates synthetic fundus images for each of the 
% 5 DR severity grades (0 to 4) for tests and pipeline validation.
%
% Inputs:
%   dataFolder        - Path to store the generated dataset (default: data/dummy)
%   numImagesPerClass - Number of images to generate per grade (default: 5)

    if nargin < 1 || isempty(dataFolder)
        srcDir = fileparts(mfilename('fullpath'));
        moduleRoot = fileparts(srcDir);
        dataFolder = fullfile(moduleRoot, 'data', 'dummy');
    end
    if nargin < 2 || isempty(numImagesPerClass)
        numImagesPerClass = 5;
    end
    
    % Seed random generator for reproducibility
    rng(42);
    
    % Create directories for each class folder (0 to 4)
    for c = 0:4
        classFolder = fullfile(dataFolder, num2str(c));
        if ~exist(classFolder, 'dir')
            mkdir(classFolder);
        end
    end
    
    % Generate images for each class
    for c = 0:4
        classFolder = fullfile(dataFolder, num2str(c));
        for i = 1:numImagesPerClass
            % Generate base black canvas
            img = zeros(256, 256, 3, 'uint8');
            
            % Compute grid coordinates and mask for circular fundus
            [X, Y] = meshgrid(1:256, 1:256);
            distFromCenter = sqrt((X - 128).^2 + (Y - 128).^2);
            fundusMask = distFromCenter <= 100;
            
            % Set base color of fundus (orange-red: [220, 100, 50])
            for ch = 1:3
                if ch == 1
                    imgVal = 220;
                elseif ch == 2
                    imgVal = 100;
                else
                    imgVal = 50;
                end
                % Add low-frequency shading representing light attenuation
                shading = 1.0 - 0.2 * (distFromCenter / 100).^2;
                imgChan = uint8(imgVal * shading .* fundusMask);
                img(:, :, ch) = imgChan;
            end
            
            % Add optic disc (warm yellow circle)
            discCenter = [128 + 40, 128 - 20];
            discMask = sqrt((X - discCenter(1)).^2 + (Y - discCenter(2)).^2) <= 15;
            discMask = discMask & fundusMask;
            imgR = img(:, :, 1); imgR(discMask) = 255; img(:, :, 1) = imgR;
            imgG = img(:, :, 2); imgG(discMask) = 255; img(:, :, 2) = imgG;
            imgB = img(:, :, 3); imgB(discMask) = 150; img(:, :, 3) = imgB;
            
            % Add blood vessels radiating from the optic disc
            for angle = 0:45:315
                rad = deg2rad(angle + randn()*10);
                xVal = discCenter(1);
                yVal = discCenter(2);
                for step = 1:80
                    xVal = xVal + cos(rad) * 1.0;
                    yVal = yVal + sin(rad) * 1.0;
                    ix = round(xVal);
                    iy = round(yVal);
                    if ix >= 1 && ix <= 256 && iy >= 1 && iy <= 256 && fundusMask(iy, ix)
                        % Draw dark red vessel pixel
                        img(iy, ix, 1) = 150;
                        img(iy, ix, 2) = 30;
                        img(iy, ix, 3) = 10;
                    end
                end
            end
            
            % Add class-specific lesions
            % Class 1: Microaneurysms (MAs) - small red dots
            if c >= 1
                numMAs = c * 5 + randi([0, 5]);
                for dot = 1:numMAs
                    rx = 128 + randi([-70, 70]);
                    ry = 128 + randi([-70, 70]);
                    if fundusMask(ry, rx)
                        img(ry-1:ry+1, rx-1:rx+1, 1) = 255;
                        img(ry-1:ry+1, rx-1:rx+1, 2) = 10;
                        img(ry-1:ry+1, rx-1:rx+1, 3) = 10;
                    end
                end
            end
            
            % Class 2: Exudates (EXs) - bright yellow blobs
            if c >= 2
                numEXs = (c - 1) * 3 + randi([0, 3]);
                for blob = 1:numEXs
                    bx = 128 + randi([-60, 60]);
                    by = 128 + randi([-60, 60]);
                    blobMask = sqrt((X - bx).^2 + (Y - by).^2) <= 3 + randi([0, 3]);
                    blobMask = blobMask & fundusMask;
                    imgR = img(:, :, 1); imgR(blobMask) = 250; img(:, :, 1) = imgR;
                    imgG = img(:, :, 2); imgG(blobMask) = 240; img(:, :, 2) = imgG;
                    imgB = img(:, :, 3); imgB(blobMask) = 180; img(:, :, 3) = imgB;
                end
            end
            
            % Class 3: Hemorrhages (HEs) - dark red blobs
            if c >= 3
                numHEs = (c - 2) * 2 + randi([0, 2]);
                for blob = 1:numHEs
                    bx = 128 + randi([-60, 60]);
                    by = 128 + randi([-60, 60]);
                    blobMask = sqrt((X - bx).^2 + (Y - by).^2) <= 6 + randi([0, 4]);
                    blobMask = blobMask & fundusMask;
                    imgR = img(:, :, 1); imgR(blobMask) = 160; img(:, :, 1) = imgR;
                    imgG = img(:, :, 2); imgG(blobMask) = 10;  img(:, :, 2) = imgG;
                    imgB = img(:, :, 3); imgB(blobMask) = 10;  img(:, :, 3) = imgB;
                end
            end
            
            % Class 4: Neovascularization (NV) - extra high-intensity branching lines
            if c >= 4
                for l = 1:3
                    bx = 128 + randi([-50, 50]);
                    by = 128 + randi([-50, 50]);
                    for step = 1:40
                        bx = bx + randn()*1.5;
                        by = by + randn()*1.5;
                        ix = round(bx);
                        iy = round(by);
                        if ix >= 2 && ix <= 255 && iy >= 2 && iy <= 255 && fundusMask(iy, ix)
                            img(iy-1:iy+1, ix-1:ix+1, 1) = 255;
                            img(iy-1:iy+1, ix-1:ix+1, 2) = 50;
                            img(iy-1:iy+1, ix-1:ix+1, 3) = 20;
                        end
                    end
                end
            end
            
            % Write image to directory
            filename = fullfile(classFolder, sprintf('img_%03d.png', i));
            imwrite(img, filename);
        end
    end
    fprintf('Successfully generated synthetic dataset under %s\n', dataFolder);
end

File: module4_dr_grading python/src/generate_dummy_dataset.py
import os
import numpy as np
import cv2

def generate_dummy_dataset(data_folder, num_images_per_class=5):
    """
    Generates synthetic fundus images for each of the 
    5 DR severity grades (0 to 4) for tests and pipeline validation.
    """
    if data_folder is None or data_folder == "":
        data_folder = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'data', 'dummy')
        
    np.random.seed(42)
    
    for c in range(5):
        class_folder = os.path.join(data_folder, str(c))
        os.makedirs(class_folder, exist_ok=True)
        
    for c in range(5):
        class_folder = os.path.join(data_folder, str(c))
        for i in range(1, num_images_per_class + 1):
            img = np.zeros((256, 256, 3), dtype=np.uint8)
            
            y, x = np.ogrid[0:256, 0:256]
            # MATLAB uses 1-indexed, here we shift slightly but it's fine.
            dist_from_center = np.sqrt((x - 127)**2 + (y - 127)**2)
            fundus_mask = dist_from_center <= 100
            
            # Base color
            R_base = 220
            G_base = 100
            B_base = 50
            
            shading = 1.0 - 0.2 * (dist_from_center / 100)**2
            
            img[:, :, 2] = np.where(fundus_mask, R_base * shading, 0) # R
            img[:, :, 1] = np.where(fundus_mask, G_base * shading, 0) # G
            img[:, :, 0] = np.where(fundus_mask, B_base * shading, 0) # B (OpenCV uses BGR)
            
            # Optic disc
            disc_center = (167, 107) # (x, y)
            disc_mask = np.sqrt((x - disc_center[0])**2 + (y - disc_center[1])**2) <= 15
            disc_mask = disc_mask & fundus_mask
            
            img[disc_mask, 2] = 255 # R
            img[disc_mask, 1] = 255 # G
            img[disc_mask, 0] = 150 # B
            
            # Blood vessels
            for angle in range(0, 316, 45):
                rad = np.deg2rad(angle + np.random.randn()*10)
                x_val = disc_center[0]
                y_val = disc_center[1]
                for step in range(80):
                    x_val += np.cos(rad) * 1.0
                    y_val += np.sin(rad) * 1.0
                    ix = int(round(x_val))
                    iy = int(round(y_val))
                    if 0 <= ix < 256 and 0 <= iy < 256 and fundus_mask[iy, ix]:
                        img[iy, ix, 2] = 150
                        img[iy, ix, 1] = 30
                        img[iy, ix, 0] = 10
                        
            # Lesions
            # Class 1: Microaneurysms
            if c >= 1:
                num_MAs = c * 5 + np.random.randint(0, 6)
                for _ in range(num_MAs):
                    rx = 127 + np.random.randint(-70, 71)
                    ry = 127 + np.random.randint(-70, 71)
                    if 1 <= ry < 255 and 1 <= rx < 255 and fundus_mask[ry, rx]:
                        img[ry-1:ry+2, rx-1:rx+2, 2] = 255
                        img[ry-1:ry+2, rx-1:rx+2, 1] = 10
                        img[ry-1:ry+2, rx-1:rx+2, 0] = 10
                        
            # Class 2: Exudates
            if c >= 2:
                num_EXs = (c - 1) * 3 + np.random.randint(0, 4)
                for _ in range(num_EXs):
                    bx = 127 + np.random.randint(-60, 61)
                    by = 127 + np.random.randint(-60, 61)
                    blob_mask = np.sqrt((x - bx)**2 + (y - by)**2) <= 3 + np.random.randint(0, 4)
                    blob_mask = blob_mask & fundus_mask
                    img[blob_mask, 2] = 250
                    img[blob_mask, 1] = 240
                    img[blob_mask, 0] = 180
                    
            # Class 3: Hemorrhages
            if c >= 3:
                num_HEs = (c - 2) * 2 + np.random.randint(0, 3)
                for _ in range(num_HEs):
                    bx = 127 + np.random.randint(-60, 61)
                    by = 127 + np.random.randint(-60, 61)
                    blob_mask = np.sqrt((x - bx)**2 + (y - by)**2) <= 6 + np.random.randint(0, 5)
                    blob_mask = blob_mask & fundus_mask
                    img[blob_mask, 2] = 160
                    img[blob_mask, 1] = 10
                    img[blob_mask, 0] = 10
                    
            # Class 4: NV
            if c >= 4:
                for _ in range(3):
                    bx = 127 + np.random.randint(-50, 51)
                    by = 127 + np.random.randint(-50, 51)
                    for step in range(40):
                        bx += np.random.randn()*1.5
                        by += np.random.randn()*1.5
                        ix = int(round(bx))
                        iy = int(round(by))
                        if 1 <= ix < 255 and 1 <= iy < 255 and fundus_mask[iy, ix]:
                            img[iy-1:iy+2, ix-1:ix+2, 2] = 255
                            img[iy-1:iy+2, ix-1:ix+2, 1] = 50
                            img[iy-1:iy+2, ix-1:ix+2, 0] = 20
                            
            filename = os.path.join(class_folder, f"patient{c}_{i:03d}_PP.png")
            cv2.imwrite(filename, img)
            
    print(f"Successfully generated synthetic dataset under {data_folder}")
